### ISSUE TITLE
Always have CSS styles overwrite inline styles

### ISSUE THREAD

**Issue by mdio** [tijsverkoyen/CssToInlineStyles#112]
We might have different use cases but for us it is important that already existing inline styles are always overwritten when the same property is in the CSS.
Right now (looking at version 2.0) they are only overwritten, if you add !important to your CSS property.

Let's say you have a button as a standard component that comes with a default look defined as inline styles and you want to give it a different background color as well as font-size.
With this you'd have to write:
```css
.my-button { 
background-color: #f00 !important;
font-size: 16px !important;
}
```
So basically you need !important everywhere.
I know that this is how CSS specificity works, but for us, doing our mails, it's not so useful.

So if there are no objections, I would love to have the "isImportant" check removed.

But maybe you have a use case for that so perhaps it should be an option instead.

**Comment by stof** [comment]
This does not make sense. It would mean that applying this library would change the rules being applied. Inline styles are winning over stylesheet rules

**Comment by mdio** [comment]
I'm aware of that.
>  I know that this is how CSS specificity works, but for us, doing our mails, it's not so useful.

It does make very much sense for us as it seems nonsensical to have a CSS file full of !important.

We have quite a lot mails and they are full of inline styles.
To be able to use this library we have three options
1. Have stylesheet rules win over inline rules
2. Have !important on every stylesheet rule
3. Migrate the whole codebase to not use inline rules at all and have stylesheets for every template/component

As said, option 1 would be preferred.
Option 2 is pretty silly and option 3 is not really practical.

**Comment by stof** [comment]
@mdio but option 1 would mean that this library does not follow the CSS spec anymore, which is a very bad idea. Build your own library if you have such custom needs IMO (and keep it internal, or make it very obvious in the readme that your library does not follow the CSS spec)

Btw, option 2 does not work: inline styles win over important styles coming from stylesheets too

**Comment by tijsverkoyen** [comment]
@mdio if I'm correct this can't be done with v1? Or am I wrong? I think *not* following the CSS spec is weird.
Wouldn't it be more sane to strip all style-attributes before passing it into the convert-method (for v2)?

**Comment by mdio** [comment]
@stof @tijsverkoyen Yes, I'm talking about version 2.0 and regarding option 2 I'm referring to [this piece of code](https://github.com/tijsverkoyen/CssToInlineStyles/blob/2.0/src/CssToInlineStyles.php#L51).

@tijsverkoyen Thanks for suggesting an alternative, but this would mean that one would have to define all styles in stylesheets if you want to override one inline style. That would boil down to option 3.

Just out of curiosity: How do you write your mails? Do you use inline styles at all? Do you only use inline styles on elements you don't style via stylesheet?

**Comment by tijsverkoyen** [comment]
Well this is how the CSS spec works, inline styles are "more important" than styles provided in a style-tag, css-file.

At SumoCoders we have a tool which will inline all the styles for us (uses this class) so we just write plain HTML and (S)CSS. Some elements are styled inline directly but we try to limit this.

**Comment by stof** [comment]
hmm, this code lets me think that the version 2.0 is broken regarding the priority of inline styles vs stylesheet styles

**Comment by tijsverkoyen** [comment]
Probably I'm missing something. @stof can you provide me with an example? And the expected output?

**Comment by stof** [comment]
@tijsverkoyen see my comment on https://github.com/tijsverkoyen/CssToInlineStyles/pull/111/files#r45736298
The removal of the existing testsuite in the refactoring is a bad idea as it removed the possibility to catch regressions in the refactoring

**Comment by mdio** [comment]
@stof I think version 2.0 gets it just right.
According to [CSS 2.2 specs](http://www.w3.org/TR/CSS21/cascade.html#specificity) inline styles have a specificity of **1,0,0,0** and according to http://cssspecificity.com/ (unfortunately, the specs are rather vague regarding !important) !important has a specificity of **1,0,0,0,0** so it should win over inline styles.

That's also how it's implemented in common browsers (you can check with this [fiddle](http://jsfiddle.net/de8oy5z2/)).

@tijsverkoyen Thanks for the insight. So we'll have to fork your project until we (if ever) come to the point where we style our mails and components solely with stylesheets.

As far as I'm concerned, this issue can be closed then.

**Comment by tijsverkoyen** [comment]
Regarding the CSS spec, I rewrote a part of the code to be respect the spec more.
Regarding this issue: I won't implement this behaviour.
